src/bindings/python/flux/cli/plugin.py

```diff
@@ -0,0 +1,260 @@
+from flux.importer import import_path
+
+
+class CLIPluginValue:
+    """Base class for dealing with values submitted through CLI. Used
+    as an argparse custom type in the base CLI implementation.
+
+    Args:
+        value (str): the user input immediately following the plugin
+            option submitted through the CLI. This value is split on
+            the '=' delimiter and expected to follow the form PLUGIN[=VAL].
+            Since the VAL is optional, self.value is True if it is not
+            provided.
+    """
+
+    def __init__(self, value):
+        temp = value.split("=")
+        self.key = temp[0]
+        try:
+            self.value = temp[1]
+        except IndexError:
+            self.value = True
+
+    def get_value(self):
+        """Getter for returning the key and value"""
+        return self.key, self.value
+
+    def __repr__(self):
+        return f"{self.key}={self.value}"
+
+
+class CLIPluginNamespace:
+    """Mutable base class for containing the value of plugins loaded in the
+    CLIPluginRegistry.
+
+    A plugin value is of the form PLUGIN[=VAL] where VAL is True if a VAL is
+    not provided. If PLUGIN is loaded but not invoked by the user, that
+    corresponding VAL is None.
+
+    The utility of this class is that it allows individual plugins to call
+    getattr(CLIPluginNamespace(), self.opt) with assurance that getattr will
+    not raise an AttributeError because all loaded plugins will have an attribute,
+    even if that attribute is None. This assurance is handled by the
+    CLIPluginRegistry().
+    """
+
+    def __init__(self):
+        pass
+
+    def __repr__(self):
+        rep = "CLIPluginNamespace("
+        for attr in dir(self):
+            if not attr.startswith("_"):
+                rep += f"{attr}={getattr(self, attr)}"
+        rep += ")"
+        return rep
+
+
+class CLIPlugin(ABC):  # pragma no cover
+    """Base class for a CLI submission plugin
+
+    A plugin should derive from this class and implement one or more
+    base methods (described below)
+
+    Attributes:
+        opt (str): command-line key passed to -P/--plugin to activate the
+            extension
+        usage (str): --help message for the extension
+        version (str): optional version for the plugin, preferred in format
+            MAJOR.MINOR.PATCH
+        prog (str): command-line subcommand for which the plugin is active,
+            e.g. "submit", "run", "alloc", "batch", "bulksubmit"
+    """
+
+    def __init__(self, prog, version=None):
+        self.prog = prog
+        if prog.startswith("flux "):
+            self.prog = prog[5:]
+        self.options = {}
+        self.version = version
+
+    def add_options(self, registry):
+        """Plugins implement this method to register options in the registry.
+
+        Plugins can call:
+        >>> registry.add_options("foo", usage="Enable foo.")
+
+        to add new options to the registry.
```

Comment:
What about instead of having a method that takes a `CLIPluginRegistry` and calls `registry.add_options`, there were a method that returned some objects with the information needed to add options? Like for this example above, return `{"name": "foo", "usage"="Enable foo."}` or something. I think passing the `registry` in to this method seems like a violation of the abstraction to me... I see there's a discussion going on a related topic that I haven't fully grok'd yet.

Reply:
It probably is a violation of abstraction. Our [`create_parser` method in `MiniCmd` is an abstract class](https://github.com/flux-framework/flux-core/blob/5f55d3c57aeed2e864e69105338668bee5461a2b/src/bindings/python/flux/cli/base.py#L757), so it has no access to `self`, which is why the parser is constructed on the fly and passed around like this. (The init method of MiniCmd sheds some light on this.)

If we we go with the alternative PR, which uses argparse's `add_option` method, plugins will have access to more than just a name and usage message, they'll have the ability to specify a ton of things, and that could get ugly. Maybe argparse provides an option class we could use for your suggestion...

Reply:
I like @jameshcorbett's suggestion. Perhaps plugins optionally implement a `get_options()` method, and the registry code queries the plugin to see what options it wants to implement. There could be a `CLIPluginOption` convenience class to allow a restricted set of argparse `add_option_argument` args and kwargs, or a well defined dict could be supported as suggested by @jameshcorbett.